sympy/functions/elementary/trigonometric.py

```diff
@@ -1887,6 +1887,11 @@ def _eval_rewrite_as_jn(self, arg, **kwargs):
     def _eval_rewrite_as_sin(self, arg, **kwargs):
         return Piecewise((sin(arg)/arg, Ne(arg, 0)), (1, True))
 
+    def _eval_rewrite_as_exp(self, arg, **kwargs):
+        a = arg.doit()
+        return Piecewise((-I*(exp(I*a) - exp(-I*a))/(2*a),
+                                              Ne(a, 0)), (1, True))
```

Comment:
I'm not sure we should call doit in a rewrite. The arg could be an Integral or something...

Can this not just be `return self.rewrite(sin).rewrite(exp)`?

Reply:
Basically it's required for the rewrite of `sinc(x**2)` to work. The inner arg `x**2` gets rewritten first to (unevaluated) `exp(2*log(x))`, then`Ne(exp(2*log(x)), 0)` automatically evaluates to `True` and the piecewise form of the expression is lost.

Reply:
I just tried with this PR and I don't seem to get a Piecewise:
```julia
In [2]: x = Symbol('x')                                                                                                           

In [3]: sin(x**2).rewrite(exp)                                                                                                    
Out[3]: 
   ⎛    2        2⎞ 
   ⎜ ⅈ⋅x     -ⅈ⋅x ⎟ 
-ⅈ⋅⎝ℯ     - ℯ     ⎠ 
────────────────────
         2          

In [4]: x = Symbol('x', positive=True)                                                                                            

In [5]: sin(x**2).rewrite(exp)                                                                                                    
Out[5]: 
   ⎛    2        2⎞ 
   ⎜ ⅈ⋅x     -ⅈ⋅x ⎟ 
-ⅈ⋅⎝ℯ     - ℯ     ⎠ 
────────────────────
         2 
```
It is generally possible to specify which parts of an expression are rewritten:
```julia
In [6]: (sin(x)+x**2).rewrite(sin, exp)                                                                                           
Out[6]: 
       ⎛ ⅈ⋅x    -ⅈ⋅x⎞
 2   ⅈ⋅⎝ℯ    - ℯ    ⎠
x  - ────────────────
            2  
```
Is the problem with the Piecewise here is that if x is posified then `Ne(x**2, 0)` will evaluate to `False`?